File: templates/problems/01_arrays_hashing/06_encode_decode_strings/encode_decode_strings_test.go
package encode_decode_strings

import (
	"reflect"
	"testing"
)

func TestEncodeDecode(t *testing.T) {
	tests := []struct {
		name string
		strs []string
	}{
		{
			name: "basic strings",
			strs: []string{"hello", "world"},
		},
		{
			name: "empty list",
			strs: []string{},
		},
		{
			name: "single empty string",
			strs: []string{""},
		},
		{
			name: "multiple empty strings",
			strs: []string{"", "", ""},
		},
		{
			name: "strings with special characters",
			strs: []string{"he:llo", "wor#ld", "foo;bar"},
		},
		{
			name: "strings with delimiters and colons",
			strs: []string{"4:abcd", "3:xyz"},
		},
		{
			name: "single character strings",
			strs: []string{"a", "b", "c"},
		},
		{
			name: "mixed empty and non-empty",
			strs: []string{"", "a", "", "b", ""},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			encoded := encode(tt.strs)
			decoded := decode(encoded)
			if decoded == nil && len(tt.strs) == 0 {
				return
			}
			if !reflect.DeepEqual(decoded, tt.strs) {
				t.Errorf("decode(encode(%v)) = %v, want %v", tt.strs, decoded, tt.strs)
			}
		})
	}
}

File: templates/problems/01_arrays_hashing/06_encode_decode_strings/typescript/encodeDecodeStrings.test.ts
import { describe, it, expect } from 'vitest';
import { encode, decode } from './encodeDecodeStrings';

describe('EncodeDecodeStrings', () => {
    it('basic strings', () => {
        const strs = ['hello', 'world'];
        expect(decode(encode(strs))).toEqual(strs);
    });

    it('empty list', () => {
        const strs: string[] = [];
        expect(decode(encode(strs))).toEqual(strs);
    });

    it('single empty string', () => {
        const strs = [''];
        expect(decode(encode(strs))).toEqual(strs);
    });

    it('multiple empty strings', () => {
        const strs = ['', '', ''];
        expect(decode(encode(strs))).toEqual(strs);
    });

    it('strings with special characters', () => {
        const strs = ['he:llo', 'wor#ld', 'foo;bar'];
        expect(decode(encode(strs))).toEqual(strs);
    });

    it('strings with delimiters and colons', () => {
        const strs = ['4:abcd', '3:xyz'];
        expect(decode(encode(strs))).toEqual(strs);
    });

    it('single character strings', () => {
        const strs = ['a', 'b', 'c'];
        expect(decode(encode(strs))).toEqual(strs);
    });

    it('mixed empty and non-empty', () => {
        const strs = ['', 'a', '', 'b', ''];
        expect(decode(encode(strs))).toEqual(strs);
    });
});
